### ISSUE TITLE
Lint `unsigned - other_unsigned > 0`

### ISSUE THREAD

**Issue by fhartwig** [Manishearth/rust-clippy#588]
Idea from [here](http://www.viva64.com/en/d/0146/).
Basically, `x - y`, where `x` and `y` are of some unsigned integer type, returns 0 iff `x == y`, so comparing the difference to 0 is dodgy and should at least be rewritten for clarity.

**Comment by mcarton** [comment]
That’d be a nice one. There is a more complete list [here](http://www.viva64.com/en/d/full/). Most of them don’t apply to Rust (yeah!), but I bet we could find other ideas there.

**Comment by fhartwig** [comment]
Yeah, I'm currently going over the list to see what might make sense for clippy. It's pretty amazing to see how many of these things are complete non-issues in rust thanks to a sensible type system and the borrow checker. There are quite a few that make sense in rust, but some of them would be tricky to implement and probably impossible to implement efficiently.

**Comment by scurest** [comment]
Would it be better to lint just `uint > 0`? And then also lint `int - int == 0`?

**Comment by fhartwig** [comment]
> Would it be better to lint just uint > 0? And then also lint int - int == 0?

I don't think there's anything wrong with `uint > 0` in general. It could be written as `!=`, but I don't think either one is strictly better than the other.

> But 4 in particular seems like it really should use == 0 instead of <= 0

Yeah, `<= 0` on unsigned types is pretty weird. I thought there was already a lint for nonsensical comparisons of `0` and unsigned types, but it looks like there isn't. We should definitely change that :smile:

**Comment by birkenfeld** [comment]
Note that `x - y > 0` is basically `{ debug_assert!(x >= y); x > y }`. The assert catches logic bugs in debug mode that just `x > y` or `x != y` doesn't, so we shouldn't just suggest replacing one by the other.
